Mathlib/Data/List/Basic.lean

```diff
@@ -197,6 +197,14 @@ theorem append_right_injective (s : List α) : Injective fun t ↦ s ++ t :=
 theorem append_left_injective (t : List α) : Injective fun s ↦ s ++ t :=
   fun _ _ ↦ append_cancel_right
 
+theorem append_cancel_right_length {as bs bs' cs : List α}
+(eq_length : bs.length = bs'.length) (h : as ++ bs = cs ++ bs') : as = cs := by
+  match as, cs with
+  | [], []       => rfl
+  | [], c::cs    => have aux := congrArg length h; simp +arith [eq_length] at aux
+  | a::as, []    => have aux := congrArg length h; simp +arith [eq_length] at aux
+  | a::as, c::cs => injection h with h₁ h₂; subst h₁; rw [append_cancel_right_length eq_length h₂]
```

Comment:
Notice that this line can also be shortened to
```suggestion
rwa [head_eq_of_cons_eq h, append_cancel_right_length _ (tail_eq_of_cons_eq h)]
```
If you'd rather have two tactics to avoid squeezing everything into a less readable line, you can also get rid of the `subst` as follows
```
injection h with h₁ h₂
rw [h₁, append_cancel_right_length eq_length h₂]
```